mypy/fastparse.py

```diff
@@ -56,20 +59,26 @@ def parse(source: Union[str, bytes], fnam: str = None, errors: Errors = None,
         else:
             ast2 = ast27.parse(source, fnam, 'exec')
             ast = conversions.py2to3(ast2)
-    except SyntaxError as e:
-        if errors:
-            errors.set_file('<input>' if fnam is None else fnam)
-            errors.report(e.lineno, e.msg)  # type: ignore
-        else:
-            raise
-    else:
+
         tree = ASTConverter(pyversion=pyversion,
                             is_stub=is_stub_file,
                             custom_typing_module=custom_typing_module,
                             ).visit(ast)
         tree.path = fnam
         tree.is_stub = is_stub_file
         return tree
+    except SyntaxError as e:
+        if errors:
+            errors.set_file('<input>' if fnam is None else fnam)
+            errors.report(e.lineno, e.msg)  # type: ignore
```

Comment:
I'm not sure what we were ignoring here before, because it works without it... except when I combine the except clauses.  It looks like then it's taking the join of the two exceptions as the type of `e` (instead of the Union) and gives me:
```
mypy/fastparse.py:73: error: "Exception" has no attribute "lineno"
mypy/fastparse.py:73: error: "Exception" has no attribute "msg"```